File: bin/app/data/src/tictactoe/tictactoe.cpp
#include "tictactoe.h"

#include "artificial_player.h"
#include "human_player.h"
#include "io.h"

TicTacToe::TicTacToe() {
	players_[0] = 0;
	players_[1] = 0;
}

TicTacToe::~TicTacToe() {
	Reset();
}

bool TicTacToe::Start() {
	Reset();
	io::stringOut("Tic Tac Toe\n\n[1] Human\n[2] Computer\n[3] Quit\n\n");

	players_[0] = SelectPlayer( Field::PlayerA, "Player A" );
	if ( !players_[0] ) {
		return false;
	}

	players_[1] = SelectPlayer( Field::PlayerB, "Player B" );
	if ( !players_[1] ) {
		return false;
	}

	io::stringOut("\n");
	return true;
}

void TicTacToe::Run() {
	field_.Show();

	int playerIndex = 0;

	for ( int i = 0; i < 9; i++ ) {
		Player& player = *players_[playerIndex];

		field_.MakeMove( player.Turn( field_ ), player.getToken() );
		field_.Show();

		if ( field_.SameInRow( player.getToken(), 3 ) ) {
			io::stringOut(player.getName());
			io::stringOut(" won!\n\n");
			return;
		}

		playerIndex = ( playerIndex + 1 ) % 2;
	}

	io::stringOut("Game ends in draw!\n\n");
}

void TicTacToe::Reset() {
	field_.Clear();

	for ( int i = 0; i < 2; i++ ) {
		if ( players_[i] ) {
			delete players_[i];
			players_[i] = 0;
		}
	}
}

Player* TicTacToe::SelectPlayer( Field::Token token, const char* name ) const {
	int selection = 0;

	while ( true ) {
		io::stringOut("Choose ");
		io::stringOut(name);
		io::stringOut(": ");

		selection = io::numberIn();

		switch ( selection ) {
			case 1 : return new HumanPlayer( token, name );
			case 2 : return new ArtificialPlayer( token, name );
			case 3 : return 0;
			default : io::stringOut("Wrong input!\n");
		}
	}
}

File: web/blog/media/prototype_one/data/tictactoe.js
data.push( { type : 'file', name : 'tictactoe.h', content : [








	{ line : 10, type : 'class', name : 'TicTacToe', content : {
		public : [
			{ line : 12, type : 'method', name : 'TicTacToe', content : [
				{ line : 13, type : 'use', member : 'players_' },
				{ line : 14, type : 'use', member : 'players_' }
			]},

			{ line : 17, type : 'method', name : '~TicTacToe', content : [
				{ line : 18, type : 'call', method : 'Reset' }
			]},

			{ line : 21, type : 'method', name : 'Start', return : 'bool', content : [
				{ line : 22, type : 'call', method : 'Reset' },
				{ line : 23, type : 'use', class : 'std::cout' },

				{ line : 25, type : 'use', member : 'players_' }, { line : 25, type : 'call', method : 'SelectPlayer' }, { line : 25, type : 'use', class : 'Field::Token', member : 'PlayerA' },
				{ line : 26, type : 'use', member : 'players_' },



				{ line : 30, type : 'use', member : 'players_' }, { line : 30, type : 'call', method : 'SelectPlayer' }, { line : 30, type : 'use', class : 'Field::Token', member : 'PlayerB' },
				{ line : 31, type : 'use', member : 'players_' },



				{ line : 35, type : 'use', class : 'std::cout' },

			]},

			{ line : 39, type : 'method', name : 'Run', content : [
				{ line : 40, type : 'use', member : 'field_' }, { line : 40, type : 'call', method : 'Show', class : 'Field' },




				{ line : 45, type : 'use', member : 'players_' },

				{ line : 47, type : 'use', member : 'field_' }, { line : 47, type : 'call', method : 'MakeMove', class : 'Field' }, { line : 47, type : 'call', method : 'Turn', class : 'Player' }, { line : 47, type : 'use', member : 'token_', class : 'Player' },
				{ line : 48, type : 'use', member : 'field_' }, { line : 48, type : 'call', method : 'Show', class : 'Field' },

				{ line : 50, type : 'use', member : 'field_' }, { line : 50, type : 'call', method : 'SameInRow', class : 'Field' }, { line : 50, type : 'use', member : 'token_', class : 'Player' },
				{ line : 51, type : 'use', class : 'std::cout' }, { line : 51, type : 'use', member : 'name_', class : 'Player' },






				{ line : 58, type : 'use', class : 'std::cout' }
			]},

		], private : [
			{ line : 62, type : 'method', name : 'Reset', content : [
				{ line : 63, type : 'use', member : 'field_' }, { line : 63, type : 'call', method : 'Clear', class : 'Field' },


				{ line : 66, type : 'use', member : 'players_' },
				{ line : 67, type : 'use', member : 'players_' },
				{ line : 68, type : 'use', member : 'players_' }


			]},

			{ line : 73, type : 'method', name : 'SelectPlayer', return : 'Player', params : [ 'Field::Token', 'std::string' ], content : [



				{ line : 77, type : 'use', class : 'std::cout' },
				{ line : 78, type : 'call', function : 'input::number' },


				{ line : 81, type : 'call', method : 'HumanPlayer', class : 'HumanPlayer' },
				{ line : 82, type : 'call', method : 'ArtificialPlayer', class : 'ArtificialPlayer' },

				{ line : 84, type : 'use', class : 'std::cout' }


			]},

			{ line : 89, type : 'member', name : 'players_', class : 'Player' },
			{ line : 90, type : 'member', name : 'field_', class : 'Field' }
		]}
	}
]});